### ISSUE TITLE
process queue cannot handle a large number of events in Google Chrome

### ISSUE THREAD

**Issue by jtarchie** [elm-lang/core#916]
tl;dr - Chrome doesn't handle `Array#shift` in an [efficient manner](https://bugs.chromium.org/p/v8/issues/detail?id=3059) for large arrays. Causing `workerQueue` to not handle lots of events in a quick manner.

Our application uses [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource to stream output from an external process. The `EventSource` is modeled as a [subscription](https://github.com/concourse/atc/blob/67b3c54b2d6ec7f95780398b834438faf1b9406e/web/elm/src/Concourse/BuildEvents.elm#L44), with each event triggers a new message.

On some streams, we can received hundreds of thousands of events, which creates a new process for each event on the elm process queue. On Google Chrome, the process queue does not get processed in a timely manner. On Firefox, however, it is near instantaneous. We were able to find with the Performance developer tool in Chrome that a lot of time is spent in `work`.

We've been able to find that `Array.shift` in Javascript is [inefficient in Google Chrome](https://bugs.chromium.org/p/v8/issues/detail?id=3059) for large arrays. This causes [`workQueue`](https://github.com/elm-lang/core/blob/f064b696a2b9ca93abf4c7d1e21ce9e35fbb1d0f/src/Elm/Kernel/Scheduler.js#L254) to choke on a large amount of processes.

With @zwilias help there is a [SSCCE example](https://ellie-app.com/scYBqsySGa1/1). It doesn't use `EventSource`, but does demonstration the slow processing of a large number of processes. Try it in Chrome and then FF. Such wow, amaze!

**Comment by zwilias** [comment]
For reference, a fairly simple fifo queue implementation that only has `enqueue` and `dequeue` operations like [this one](https://github.com/crash83k/advanced-queue/blob/master/lib/queue.js) make performance on chrome manageable. (It basically cheats by acting as if dequeued elements are no longer in the array, periodically slicing the start off in a batch.)
